### PR TITLE
FOGL-8750 : ReadingSet Circular Buffer implemented

### PR DESCRIPTION
**Pull request by gnandan**
Unit test result
```
./RunTests

Repeating all tests (iteration 1) . . .

Note: Randomizing tests' orders with a seed of 38991 .
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from TESTCircularBuffer
[ RUN      ] TESTCircularBuffer.TestMaxLimitOfBuffer
[       OK ] TESTCircularBuffer.TestMaxLimitOfBuffer (1 ms)
[ RUN      ] TESTCircularBuffer.TestCustomSizeBufferLessThanOne
[       OK ] TESTCircularBuffer.TestCustomSizeBufferLessThanOne (0 ms)
[ RUN      ] TESTCircularBuffer.TestHeadAndTailMarkerAdjustment
[       OK ] TESTCircularBuffer.TestHeadAndTailMarkerAdjustment (0 ms)
[ RUN      ] TESTCircularBuffer.TestCustomSizeBuffer
[       OK ] TESTCircularBuffer.TestCustomSizeBuffer (0 ms)
[ RUN      ] TESTCircularBuffer.TestExtactFromEmptyBuffer
[       OK ] TESTCircularBuffer.TestExtactFromEmptyBuffer (0 ms)
[----------] 5 tests from TESTCircularBuffer (6 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (8 ms total)
[  PASSED  ] 5 tests.
```

### CODE REVIEW

**Comment by AmandeepArora** [comment]
Valgrind tests have been run?

File: C/common/readingset_circularbuffer.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ReadingSetCircularBuffer::ReadingSetCircularBuffer(unsigned long maxBufferSize) : m_maxBufferSize(maxBufferSize), m_head(0), m_tail(-1)
+{
+}
+
+/**
+ * Destructor for a result set
+ */
+ReadingSetCircularBuffer::~ReadingSetCircularBuffer()
+{
+	/* Delete the readings */
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:29]: mutex needs to be acquired here also.

File: C/common/readingset_circularbuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		if (m_head > m_tail)
+             m_head--;
+	}	
+
+	std::vector<Reading *> *newReadings = new std::vector<Reading *>;
+	
+	// Iterate over all the readings in ReadingSet
+	for (auto const &reading : readings)
+	{
+		std::string assetName = reading->getAssetName();
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:77]: You can probably use Reading class's copy constructor directly here.

File: C/common/readingset_circularbuffer.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+std::vector<std::shared_ptr<ReadingSet>> ReadingSetCircularBuffer::extract(bool isExtractSingleElement)
+{
+	lock_guard<mutex> guard(m_mutex);
+	bool isEmpty = isBufferEmpty();
+	
+	if (isEmpty)
+		 Logger::getLogger()->info("ReadingSet circular buffer is empty");
+
+	if (!isExtractSingleElement || isEmpty)
+		return m_circularBuffer;
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:147]: Need to increment head/nextReadIdx in this case also.

File: tests/unit/C/common/test_circular_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    //First ReadingSet
+    vector<Reading *> *readings1 = new vector<Reading *>;
+    long dpVal1 = 30;
+    DatapointValue dpv1(dpVal1);
+    readings1->emplace_back(new Reading("R1", new Datapoint("DP1", dpv1)));
+    ReadingSet* rs1 = new  ReadingSet(readings1);
+    buffer.insert(rs1);
+    delete readings1;
+    std::vector<std::shared_ptr<ReadingSet>> buff1 = buffer.extract();
+    ASSERT_EQ(buff1.size(),1);
```
Comment (AmandeepArora) [inline tests/unit/C/common/test_circular_buffer.cpp:216]: May be also compare the reading returned in buff1 with readings1, fieldwise, may be by using toJSON() and doing string comparison.

File: C/common/readingset_circularbuffer.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	
+	std::vector<std::shared_ptr<ReadingSet>> bufferedItem;
+	bufferedItem.emplace_back(m_circularBuffer[m_head]);
+	
+	if (m_head  < m_circularBuffer.size())
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:152]: If we have returned all ReadingSets from the buffer, we shouldn't return any more ReadingSets on subsequent extract operations.

File: C/common/readingset_circularbuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+/**
+ * Check if circular buffer is empty
+ *
+ * @return	Return true if circular buffer is empty otherwise false
+ *
+ */
+bool ReadingSetCircularBuffer::isBufferEmpty()
+{
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:165]: Add mutex check here.

File: C/common/readingset_circularbuffer.cpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+/**
+ * Check if circular buffer is full
+ *
+ * @return	Return true if circular buffer is full otherwise false
+ *
+ */
+bool ReadingSetCircularBuffer::isBufferFull()
+{
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:176]: Add mutex check here.

File: C/common/readingset_circularbuffer.cpp [hunk 8]
```diff
@@ -57,75 +65,26 @@ void ReadingSetCircularBuffer::insert(ReadingSet* readings)
...
+	bool isBufferFull = isFull();
 	lock_guard<mutex> guard(m_mutex);
 	//Check if there is space available to insert a new ReadingSet
-	if (isBufferFull())
+	if (isBufferFull)
 	{
 		Logger::getLogger()->info("ReadingSetCircularBuffer buffer is full, removing first element");
-		// Make space for new ReadingSet and adjust m_head marker
+		// Make space for new ReadingSet and adjust m_nextReadIndex
 		m_circularBuffer.erase(m_circularBuffer.begin() + 0);
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:75]: remove "+0", if unnecessary.

File: C/common/readingset_circularbuffer.cpp [hunk 9]
```diff
@@ -57,75 +65,26 @@ void ReadingSetCircularBuffer::insert(ReadingSet* readings)
  */
 void ReadingSetCircularBuffer::appendReadingSet(const std::vector<Reading *>& readings)
 {
+	bool isBufferFull = isFull();
 	lock_guard<mutex> guard(m_mutex);
 	//Check if there is space available to insert a new ReadingSet
-	if (isBufferFull())
+	if (isBufferFull)
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:71]: The buffer fullness status can change between isFull() call above and current line in multi-threading setting.

File: C/common/readingset_circularbuffer.cpp [hunk 10]
```diff
@@ -137,22 +96,37 @@ void ReadingSetCircularBuffer::appendReadingSet(const std::vector<Reading *>& re
  */
 std::vector<std::shared_ptr<ReadingSet>> ReadingSetCircularBuffer::extract(bool isExtractSingleElement)
 {
-	lock_guard<mutex> guard(m_mutex);
-	bool isEmpty = isBufferEmpty();
 	
-	if (isEmpty)
-		 Logger::getLogger()->info("ReadingSet circular buffer is empty");
+    bool isBufferEmpty = isEmpty();
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:100]: Buffer empty state can change between lines 100 and 103.

File: C/common/readingset_circularbuffer.cpp [hunk 11]
```diff
@@ -161,9 +135,10 @@ std::vector<std::shared_ptr<ReadingSet>> ReadingSetCircularBuffer::extract(bool
  * @return	Return true if circular buffer is empty otherwise false
  *
  */
-bool ReadingSetCircularBuffer::isBufferEmpty()
+bool ReadingSetCircularBuffer::isEmpty()
 {
-	return m_circularBuffer.empty();
+	lock_guard<mutex> guard(m_mutex);
+	return m_circularBuffer.empty() || (m_nextReadIndex == m_circularBuffer.size());
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:141]: This function basically returns whether there are new items that can be read/extracted based on m_nextReadIndex, so either rename this function appropriately or change the logic here.

File: C/common/include/readingset_circularbuffer.h [hunk 12]
```diff
@@ -33,7 +33,7 @@ class ReadingSetCircularBuffer {
 		
 
 	private:
-		std::mutex	m_mutex;
+		std::recursive_mutex	m_mutex;
```
Comment (AmandeepArora) [inline C/common/include/readingset_circularbuffer.h:36]: I am not sure if it is a good idea to use recursive_mutex everywhere and let mutex be acquired multiple times recursively even in a single use-case flow. It is probably better to just "check for buffer full and handle overflow" in append() and "check for buffer empty and return empty ReadingSet vector" in extract() in a single function and remove checkFull/checkEmpty functions.

File: C/common/readingset_circularbuffer.cpp [hunk 13]
```diff
@@ -98,8 +98,8 @@ void ReadingSetCircularBuffer::appendReadingSet(const std::vector<Reading *>& re
 std::vector<std::shared_ptr<ReadingSet>> ReadingSetCircularBuffer::extract(bool isExtractSingleElement)
 {
 	
-    std::lock_guard<std::recursive_mutex> guard(m_mutex);
-    bool isUnreadDataInBuffer = isEmpty() || (m_nextReadIndex == m_circularBuffer.size());
+    lock_guard<mutex> guard(m_mutex);
+    bool isUnreadDataInBuffer = m_circularBuffer.empty() || (m_nextReadIndex == m_circularBuffer.size());
```
Comment (AmandeepArora) [inline C/common/readingset_circularbuffer.cpp:102]: Your variable name has the opposite meaning. You are checking whether m_circularBuffer doesn't have unread data. May be invert the boolean condition on the right hand side or rename the variable.

**Review by AmandeepArora — APPROVED** [review]

**Review by AmandeepArora — DISMISSED** [review]

**Review by MarkRiddoch — COMMENTED** [review]
Just a few minor comments

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**